### PR TITLE
While deleting a namespace, the TPR instances under this ns should be…

### LINKED ISSUE #46736: TPR objects still exist after deleting namespaces
**BUG REPORT**

**Kubernetes version** (use `kubectl version`): v1.6.2

**Environment**:
- **Cloud provider or hardware configuration**: GKE

**What happened**:
- Create TPR "MyResource"
- Create namespace "soak-test"
- Create a "MyResource" "a" under "soak-test":

```
$ cat my-resource-a.yaml
apiVersion: "example.com/v1"
kind: "MyResource"
metadata:
  name: "a"
...
$ kubectl --namespace soak-test create -f my-resource-a.yaml
$ kubectl --namespace soak-test get MyResource
NAME           LABELS    DATA
a           ...
```
- Delete namespace "soak-test"

```
$ kubectl delete ns soak-test
```
- wait namespace to be deleted

MyResource "a" still exists after ns is deleted.
```
$ kubectl --namespace soak-test get MyResource
NAMESPACE   NAME                          KIND
soak-test   a
```

### PR DESCRIPTION
**Pull request by xingzhou**
… deleted.

While deleting a namespace, the TPR instances under this ns should be deleted.

Fixed #46736 

**Release note**:
```
None
```

### CODE REVIEW

**Comment by jdef** [comment]
@k8s-bot ok to test

File: cmd/kube-controller-manager/app/core.go [hunk 1]
```diff
@@ -123,7 +123,15 @@ func startNamespaceController(ctx ControllerContext) (bool, error) {
 		return true, fmt.Errorf("failed to parse preferred server resources: %v", err)
 	}
 	discoverResourcesFn := namespaceKubeClient.Discovery().ServerPreferredNamespacedResources
-	if _, found := gvrs[extensions.SchemeGroupVersion.WithResource("thirdpartyresource")]; !found {
+	// Find out whether TPR is enabled despite the version
+	tprEnabled := false
+	for gvr := range gvrs {
+		if gvr.Group == extensions.GroupName && gvr.Resource == "thirdpartyresources" {
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: I'm against any special casing for TPRs.  Just make the discovery always dynamic.

Comment (xingzhou) [inline cmd/kube-controller-manager/app/core.go]: sure, let me update the patch

**Comment by xingzhou** [comment]
Removed the static discovery part in the latest patch, please take a look, thanks

**Comment by nikhita** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by nikhita** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by xingzhou** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Pull request merged**